### ISSUE TITLE
Add instructions for running unit tests and end-to-end tests in README.rst

### ISSUE THREAD

**Issue by sunrisesarsaparilla** [renpy/renpy#5808]
As a contributor to Ren'Py, unit tests and end-to-end tests can help me avoid introducing regressions when making contributions. However, I don't know how to execute either type of test. Adding instructions in the README.rst for how to execute both would be appreciated.

**Comment by renpytom** [comment]
We don't really have those at the moment.

**Comment by sunrisesarsaparilla** [comment]
Follow-up questions:

Is the code in the following locations dead code, then? 

https://github.com/renpy/renpy/tree/fca89d813ec176db56bc7db9d4d27be03b67540a/unittests
https://github.com/renpy/renpy/tree/fca89d813ec176db56bc7db9d4d27be03b67540a/testcases/game
https://github.com/renpy/renpy/tree/fca89d813ec176db56bc7db9d4d27be03b67540a/renpy/test

Also, does this mean that you manually test changes? I feel concerned about how burdensome that might be. Would you like to have end-to-end and unit tests?

**Comment by renpytom** [comment]
So, there was a testcase framework. It has't been kept up to date, though there are some, improvements in a PR or branch. 

I do mostly manually test changes. The problem with a game engine is the sheer combinatorial explosions of settings and so on that make automatic testing hard.

There's also a general sense that writing testcases feels a lot like work, and since Ren'Py is a project I do after work, I don't want it to feel like work.

**Comment by sunrisesarsaparilla** [comment]
Would you like unit tests and a few basic end-to-end tests if others add them?

**Comment by renpytom** [comment]
I'm more interested in functional tests than unit tests. The amount of test harnessing for unit tests seems like the effort wouldn't be worth it for all but the simplest function. By contrasts, renpy.test had a lot there, and @Gouvernathor  did improve it a bit - so perhaps the effort to get it working would be worth it.

**Comment by mathrick** [comment]
> I do mostly manually test changes. The problem with a game engine is the sheer combinatorial explosions of settings and so on that make automatic testing hard.

While I do sympathise with the sentiment (I've been there myself, staring at the immense amount of work needed to add tests to an existing project, and I know it's genuinely hard to appreciate the benefits of a robust test suite if you've never had one in place), it's also completely misguided. Unit tests are absolutely indispensable in any project larger than a throwaway one-liner, and the larger the project, the more they are worth the upfront investment.

It's precisely the combinatorial explosion of settings that makes manual testing a non-starter, and necessitates automated unit testing wherever possible. Integration and functional tests are a necessary complement to the unit tests, but they are not a replacement, and in either case, if it's not automated, it might as well not be there. The whole point of having an automated unit test suite is to ensure that each component has a well-defined, well-understood and stable external behaviour that other components can rely on, and that changes to its behaviour won't go unnoticed. Because you're testing individual components in isolation, the combinatorial explosion doesn't happen, and as a bonus, any under- or ill-specified aspects of it will likely be caught and hopefully documented. Once you can trust that your components are behaving as they're supposed to, the role of functional and integration tests becomes merely verifying that they remain *compatible* (ie. nothing got forgotten when making an intentional change), rather than ensuring they are individually functional. Thus once again, they grow much slower than the combinatorial explosion.

Yes, it's a lot of work to add a full test suite to an existing, large project, but the good news is, you don't need to have a full suite for it to be useful. Any area you add tests to will immediately become more stable, better defined and easier to work on with confidence that things aren't simultaneously breaking. Once you have the skeleton in place, you can add tests to components one by one, and their value will grow with each added test. The only alternative to fully automated testing is, unfortunately, a brittle sprawling mess that's near impossible to extend without breaking things elsewhere, and that is sadly showing in Ren'Py. I have a moderate size project with very few custom components, but already the few custom functions that it has are running into what's most likely bugs and not intended behaviour (hard to tell without tests to confirm), and on top of that, they can't even be upgraded away from 8.0.3 because everything from 8.1.x onwards behaves entirely differently (again, presumably not intentionally so, since I can't find anything in the docs that would suggest the intent to change). It might not feel like it right now, but believe me, having tests in place is *less* work in the long run, by a long shot.

**Comment by renpytom** [comment]
@mathrick Are you planning to add tests?

**Comment by mathrick** [comment]
@renpytom: I don't know the internals well enough to add much in terms of individual tests, but if you need help adding the groundwork for testing and getting started with how to write and utilise tests, then that is something I'm very qualified to do. I can help you put the overall infrastructure in place, as well document it and ensure it runs automatically as appropriate, i.e. before pull requests are merged, before release, etc. I've worked with GitHub Actions previously, so this is something I'd be able to help you set up. I have a moderate amount of free time at the moment, so if you're interested, I could redirect some of it towards adding tests to Ren'Py.

**Comment by renpytom** [comment]
I don't think the infrastructure is the hard part - it's the work to create this series of tests. We already have quite a bit of infrastructure in place, but without people that want tests making the commitment to update them, the infrastructure is likely to go unused. 

I played around with unit testing (and even functional testing), and my experience was that it it didn't find enough issues to feel like it was a good use of time, compared to everything else I can spend time on. Functional testing was better, but in general issues in Ren'Py tend to be caused by intentional behavior changes that have unexpected side effects, and that's something that unit tests have trouble catching - since the new behavior is what's intended.

**Comment by sunrisesarsaparilla** [comment]
I created a _very rough_ end-to-end test to check if a game's build artifacts had changed, so that I could detect if I introduced any regressions when refactoring Ren'Py code. I can't gauge how useful such a test would be for others, though, so I'd like to hear thoughts on this.

I can imagine it being useful because:
- it's as high level as can be, and thus robust

I can imagine it not being useful because:
- the test wouldn't work whenever the files that are included in the builds themselves (such as the Ren'Py engine) are changed

Additionally, even if this would be useful, there are few challenges that I ran into when creating this test that would need to be addressed:
1. Since Ren'Py builds aren't deterministic, I had to replace a few lines of code with constants (for example, replacing a call to `time.time()` with `0`). This was just an experiment, though, and those changes probably broke something. I wonder if there's a non-breaking way to make those code lines deterministic (such as hashing). I don't know how hard this would be to change.
2. The compiled python code wasn't bitwise identical, because the version of Python that Ren'Py uses embeds timestamps in files. Adding a compilation flag to use hashing instead of timestamps mostly resolved this problem. However, the version of Python that Ren'Py uses has a bug where even with hashing, certain conditions aren't deterministically compiled, which affected a handful of Ren'Py files. I think this bug was fixed in a later version of Python, however, and it also uses hashes by default.

Would a test like this be helpful?

**Comment by renpytom** [comment]
@sunrisesarsaparilla 

I'm not sure what the benefits of that test would be. Just about every change to Ren'Py will affect some aspect of the build game, even if just because the Python that makes up Ren'Py has  changed. What sort of changes would make it such that the test failing is interesting to us? 

I think I'm misunderstanding what the end-to-end build test does.

**Comment by sunrisesarsaparilla** [comment]
@renpytom

I created the test to check against regressions while refactoring, and only while refactoring, renpy code that wasn't part of the renpy engine itself. I created the test when I was working on introducing parallelization to game builds, to address this issue: https://github.com/renpy/renpy/issues/2420 I figured such a test was critical to make sure that any changes didn't mess up the game build process.

However, I don't know how often code outside of the engine itself gets refactored, so I can't gauge whether this test would be significantly beneficial in other use cases, and thus worthwhile. It sounds like you think it wouldn't be significantly useful, if I understand you.

**Comment by renpytom** [comment]
I'm not going to spend months and months writing test cases, so unless someone steps up to create a working test suite, I don't see this going very far. I'm not going to be argued to about how I choose to spend my time. 

It's generally been my experience that unit tests have been productive in the context of Ren'Py. That's because like I said, the problem is rarely a straight up bug. More likely is that a function has some sort of behavior change, because it's behavior was either wrong or incomplete, and that intended change affects unrelated functions - beyond the unit. 

What I do think would be a much better testing approach would be automatic functional tests - running test games with canned inputs, and making sure the game's outputs matches what was expected. 

If you have a good idea of how to use unit tests to verify, for example, that ATL behavior remains the same between revisions, then I'd be willing to entertain that. But I don't see unit tests as being the way to do that.

**Comment by sunrisesarsaparilla** [comment]
Since this conversation is long, here's a summary for future reference:

Unit testing in Ren'Py has a low ROI because VNs have a _lot_ of state, which is intrinsic to VNs, which results in a combinatorially impossibly large quantity of state to test. Plus, a lot of people don't want to write tests anyway, and this is a volunteer project, so what someone wants to do matters quite a bit.

Automatic functional tests might be useful, however, if someone wants to invest time into making them. Such tests would be "running test games with canned inputs, and making sure the game's outputs matches what was expected." For example, a test might simulate a user making choices in a game, and check if the user is on the right screen.

**Comment by renpytom** [comment]
@sunrisesarsaparilla 

That's about right. 

One thing I'll add is that Ren'Py is a combination programming language and game engine. It's useless without a game script, and that makes testing it a bit different than testing other software, which is at least present as a complete unit. 

There is a testcase sub language in Ren'Py. It's a bit out of date, but one thing I do want is to make it useful again. (It doesn't handle visual results, though, which are the hardest thing to avoid changes in. Maybe that can be fixed without too much work.)

**Comment by Drachenfels** [comment]
I wanted to contribute to the discussion. I am writing a visual novel, and the lack of a test suite for that game is a major headache. I am considering right now to play a bit with pytest which somehow sends keyboard/mouse inputs to the game and then observes captured logs and/or changes on the screen by capturing output and running image recognition - I did something like that in the past for smart tv's testing. It's pretty straightforward on Linux, but having some support for this in renpy (especially for automating controls and state manipulation) would help immensely. 

TLDR; I would support anyone who attempts this.

In my closing words, I will put a quote from a recent Factorio dev-blog:

>I want to restate, that without our almost 6,000 automated tests, we would never be able to avoid reintroducing old bugs by fixing new bugs, and we hope, that it gets more recognized, and gaming companies, will recognize this approach as the gold standard.

And video of it in action is truly the work of art:

https://www.youtube.com/watch?v=CgMV2dFFdFE
